samples: nrf9160: lwm2m_client:  librarization of the common parts

VeijoPesonen opened the pull request:
The idea here is to introduce a an nRF91 specific LwM2M client utility library which initializes LwM2M objects found from almost all LwM2M enabled applications. The library defines following objects:
* Device object
* Security object
* Modem Connection Monitoring object
* Firmware object
* Location object

Additionally this PR is supposed to introduce documentation which explains what does the library offer and how a user may add new objects and resources to his application. It's assumed that users might use the already existing nRF9160: LwM2M Client sample as basis for their own application and their desire is to have proper documentation which tell show to expand it with new objects, resources etc.

FOTA buff alignment issue got solved when the undersigned accidentally happened to check PR #4604. The buffer needs to be word aligned due to PR #4574.

VeijoPesonen commented on samples/nrf9160/lwm2m_client/Kconfig:
```diff
@@ -6,10 +6,9 @@
-
 config APP_LWM2M_SERVER
 	string "LwM2M server name or IP address"
+	depends on LWM2M_CLIENT_LIB
```
How this should be done correctly?

VeijoPesonen replied:
Removed as unnecessary.

rlubos commented:
> How should I split the different configuration options listed in prj.conf of the sample. Some of those shall be left there but which should be placed in lib/lwm2m_client/Kconfig?

Not sure I follow this question. You don't move configuration from prj.conf into Kconfig file. Unless you meant to move options from the sample Kconfig file into the other one. Then with such a rework practically everything should be moved.

Anyway, I'd like to understand the reasoning behind these changes first. Why are you splitting the sample? Who would be the user of the library? I kind of fail to see how these sample-specific (and nRF9160 specific) object definitions could be useful in a more generic library (maybe with some exceptions).

tejlmand commented on lib/lwm2m_client/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+zephyr_include_directories(include)
+zephyr_library()
+
+file(GLOB FOTA_SRC "${CMAKE_CURRENT_LIST_DIR}/fota/*.c")
+file(GLOB LWM2M_SRC "${CMAKE_CURRENT_LIST_DIR}/lwm2m/*.c")
+file(GLOB UI_SRC "${CMAKE_CURRENT_LIST_DIR}/ui/*.c")
```
There is no reason for using a glob pattern here.
Please use:
```
zephyr_library_sources(<files>)
```

That will also allow you to do:
```
zephyr_library_sources_ifdef(...)
```

```suggestion
```

VeijoPesonen replied:
Fixed.

tejlmand commented on lib/lwm2m_client/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+file(GLOB LWM2M_SRC "${CMAKE_CURRENT_LIST_DIR}/lwm2m/*.c")
+file(GLOB UI_SRC "${CMAKE_CURRENT_LIST_DIR}/ui/*.c")
+
+if(NOT CONFIG_LWM2M_LOCATION_OBJ_SUPPORT)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_location.c")
+endif()
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_LWM2M_LOCATION_OBJ_SUPPORT lwm2m/lwm2m_location.c)
```

VeijoPesonen replied:
Fixed.

tejlmand commented on lib/lwm2m_client/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+if(NOT CONFIG_LWM2M_LOCATION_OBJ_SUPPORT)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_location.c")
+endif()
+if(NOT CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_firmware.c")
+endif()
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT lwm2m/lwm2m_firmware.c)

VeijoPesonen replied:
Fixed.

tejlmand commented on lib/lwm2m_client/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+if(NOT CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_firmware.c")
+endif()
+if(NOT CONFIG_LWM2M_CONN_MON_OBJ_SUPPORT)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_connmon.c")
+endif()
+if(NOT CONFIG_LWM2M_IPSO_LIGHT_CONTROL)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_light_control.c")
+endif()
+if(NOT CONFIG_LWM2M_IPSO_TEMP_SENSOR)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_temp.c")
+endif()
+if(NOT CONFIG_UI_BUZZER)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_buzzer.c")
+endif()
+if(NOT CONFIG_UI_BUTTON)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_button.c")
+endif()
+if(NOT CONFIG_LWM2M_IPSO_ACCELEROMETER)
+	list(REMOVE_ITEM LWM2M_SRC "${CMAKE_CURRENT_SOURCE_DIR}/lwm2m/lwm2m_accelerometer.c")
+endif()
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_LWM2M_CONN_MON_OBJ_SUPPORT lwm2m/lwm2m_connmon.c)
zephyr_library_sources_ifdef(CONFIG_LWM2M_IPSO_LIGHT_CONTROL   lwm2m/lwm2m_light_control.c)
zephyr_library_sources_ifdef(CONFIG_LWM2M_IPSO_TEMP_SENSOR     lwm2m/lwm2m_temp.c)
zephyr_library_sources_ifdef(CONFIG_UI_BUZZER                  lwm2m/lwm2m_buzzer.c)
zephyr_library_sources_ifdef(CONFIG_UI_BUTTON                  lwm2m/lwm2m_button.c)
zephyr_library_sources_ifdef(CONFIG_LWM2M_IPSO_ACCELEROMETER   lwm2m/lwm2m_accelerometer.c)
```

Remember to add any non-covered files, that are no longer added when globbing is not used.

VeijoPesonen replied:
Fixed.

tejlmand commented on lib/lwm2m_client/CMakeLists.txt:
```diff
@@ -0,0 +1,52 @@
+endif()
+
+
+if(NOT CONFIG_UI_BUTTON)
+  list(REMOVE_ITEM UI_SRC ${CMAKE_CURRENT_SOURCE_DIR}/ui/button.c)
+endif()
+if(NOT CONFIG_UI_BUZZER)
+  list(REMOVE_ITEM UI_SRC ${CMAKE_CURRENT_SOURCE_DIR}/ui/buzzer.c)
+endif()
+if(NOT CONFIG_UI_LED_USE_PWM)
+  list(REMOVE_ITEM UI_SRC ${CMAKE_CURRENT_SOURCE_DIR}/ui/led_pwm.c)
+endif()
```
```suggestion
zephyr_library_sources_ifdef(CONFIG_UI_BUTTON      ui/button.c)
zephyr_library_sources_ifdef(CONFIG_UI_BUZZER      ui/buzzer.c)
zephyr_library_sources_ifdef(CONFIG_UI_LED_USE_PWM ui/led_pwm.c)
```

VeijoPesonen replied:
Fixed.

tejlmand commented on lib/lwm2m_client/Kconfig:
```diff
@@ -0,0 +1,20 @@
+module-str = LWM2M Client utilities
+source "${ZEPHYR_BASE}/subsys/logging/Kconfig.template.log_config"
+
+endif # LWM2M_CLIENT_LIB
```
missing new line

VeijoPesonen replied:
Fixed.

tejlmand reviewed (changes requested):
Please remove the globbing expressions.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-4257
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

rlubos reviewed (commented):
Looks way better now. We could also consider adding individual Kconfigs for each default object implementation so that they could easily be overridden by app-specific implementation, but I guess for startes a single Kconfig enabling all of the objects should do the job as well.

VeijoPesonen commented on include/net/nrf91_lwm2m_app.h:
```diff
@@ -0,0 +1,42 @@
+int lwm2m_init_device(void);
+
+/* Location */
+int lwm2m_init_location(void);
```
Fixed.

VeijoPesonen commented on include/net/nrf91_lwm2m_app.h:
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
Fixed.

VeijoPesonen commented on lib/nrf91_lwm2m_app/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor
```
Fixed.

VeijoPesonen commented on lib/nrf91_lwm2m_app/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+
+zephyr_library()
+
+zephyr_library_sources(fota/settings.c)
```
Fixed.

VeijoPesonen commented on lib/nrf91_lwm2m_app/lwm2m/lwm2m_device.c:
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
Fixed.

VeijoPesonen commented:
> 
> 
> Looks way better now. We could also consider adding individual Kconfigs for each default object implementation so that they could easily be overridden by app-specific implementation, but I guess for startes a single Kconfig enabling all of the objects should do the job as well.

Ok, that's something I should maybe still do as part of this PR.

VeijoPesonen commented:
I made the Kconfig adjustments on the library side. Next thing is to address the sample itself.
![image](https://user-images.githubusercontent.com/33285290/114995350-64edd500-9ea6-11eb-9fc8-6a37bed8b3df.png)

VeijoPesonen commented:
I did go throught the samples Kconfig.

![image](https://user-images.githubusercontent.com/33285290/115046971-bca93200-9ee0-11eb-9ce4-5c3d11107dec.png)

![image](https://user-images.githubusercontent.com/33285290/115047109-e4989580-9ee0-11eb-98ae-67bb810ba317.png)

rlubos commented on lib/nrf91_lwm2m_app/Kconfig:
```diff
@@ -0,0 +1,98 @@
+endif #COAP_EXTENDED_OPTIONS_LEN
+endmenu
+
+config LWM2M_SERVER_DEFAULT_PMIN
```
Actually, I don't think that configs like these should be altered by the library in any way. The `pmin`/`pmax` for instance is a core LwM2M feature and I don't think the library needs to alter this value in any way. Same goes with some other configs like JSON support or COAP ACK timeout.

VeijoPesonen replied:
Ok, will fix that one. I was also thinking should I bring any of the changes on the overlay files as part of Kconfig. I mean here the bootstrap server settings, queue mode and NB-IoT-specific settings.

rlubos replied:
I'd be really cautious with moving these configs to be honest - the library should provide the default implementation of certain objects, not enforce specific configurations. For instance the port number - it's tightly bound with the sever being used, I don't think the library should enforce any value - I'd say it's an application-specifc config (like the server URL).

In my opinion we shouldn't introduce too many configs to the library. I was rather thinking we should have configs like `CONFIG_NRF91_LWM2M_APP_FIRMWARE_UPDATE_OBJ` that would enable the Firmware Object implementation in the library, and which could `select` any other configs from the core library to make it work. The user would then have the flexibility to disable parts of the library that he wants to re-implement differently.

VeijoPesonen replied:
>  I was rather thinking we should have configs like CONFIG_NRF91_LWM2M_APP_FIRMWARE_UPDATE_OBJ that would enable the Firmware Object implementation in the library, and which could select any other configs from the core library to make it work.

That is still to be done but I hope the split of the configurations is now you hope it to be.

VeijoPesonen replied:
Hopefully fixed fully now.

VeijoPesonen commented:
Started the documentation work even though I'm not done yet what you @rlubos requested. No need to review the documentation changes yet as there are still many things to be done. 

![image](https://user-images.githubusercontent.com/33285290/115685780-39189680-a361-11eb-9e4d-eba88174a894.png)

VeijoPesonen commented on lib/nrf91_lwm2m_app/Kconfig:
```diff
@@ -0,0 +1,47 @@
+	default y
+	select LWM2M_CONN_LOCATION_OBJ_SUPPORT
+
+config NRF91_LWM2M_APP_FIRMWARE_UPDATE_OBJ_SUPPORT
```
@rlubos Are these macros something what you were looking for?

rlubos replied:
yes, thank you

VeijoPesonen commented on lib/nrf91_lwm2m_app/Kconfig:
```diff
@@ -0,0 +1,98 @@
+
+if NRF91_LWM2M_APP
+
+config NET_NATIVE
```
Not relevant anymore as I changed the configuration for something different.

rlubos commented on include/net/nrf91_lwm2m_app.h:
```diff
@@ -0,0 +1,89 @@
+ */
+int lwm2m_init_device(void);
+
+#if defined CONFIG_NRF91_LWM2M_APP_LOCATION_OBJ_SUPPORT
```
```suggestion
#if defined(CONFIG_NRF91_LWM2M_APP_LOCATION_OBJ_SUPPORT)
```

VeijoPesonen replied:
Fixed.

rlubos commented on include/net/nrf91_lwm2m_app.rst:
```diff
@@ -0,0 +1,55 @@
+nRF91 based device to an LwM2M server such as `Leshan Demo Server`_ via LTE. Once connected, the device can be queried
+to obtain location data, and to retrieve information about the modem.
+
+This library provides a basic set of LwM2M resources combined together to avoid the need to duplicate these
```
```suggestion
This library provides a basic set of LwM2M objects combined together to avoid the need to duplicate these
```

VeijoPesonen replied:
Done.

rlubos commented on include/net/nrf91_lwm2m_app.rst:
```diff
@@ -0,0 +1,55 @@
+
+This library provides a basic set of LwM2M resources combined together to avoid the need to duplicate these
+in all of the applications planning to communicate with an LwM2M server. Following fixed set of readily initialized
+resources are at users exposal.
```
```suggestion
objects are at users exposal.
```

VeijoPesonen replied:
Done.

rlubos commented on lib/nrf91_lwm2m_app/lwm2m/lwm2m_firmware.c:
```diff
@@ -16,15 +16,16 @@
 #define REBOOT_DELAY       K_SECONDS(1)
 
-#if defined(CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT)
+#if defined(CONFIG_NRF91_LWM2M_APP_FIRMWARE_UPDATE_OBJ_SUPPORT)
```
I don't think we need those ifdefs in the source code, this whole file is included conditionally based on `CONFIG_NRF91_LWM2M_APP_FIRMWARE_UPDATE_OBJ_SUPPORT`. A bit weird they were here in the first place, probably some leftover code.

VeijoPesonen replied:
Fixed.

rlubos commented on lib/nrf91_lwm2m_app/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+zephyr_library_sources_ifdef(CONFIG_NRF91_LWM2M_APP_LOCATION_OBJ_SUPPORT lwm2m/lwm2m_location.c)
+zephyr_library_sources_ifdef(CONFIG_NRF91_LWM2M_APP_FIRMWARE_UPDATE_OBJ_SUPPORT lwm2m/lwm2m_firmware.c)
+zephyr_library_sources_ifdef(CONFIG_NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT lwm2m/lwm2m_connmon.c)
+zephyr_library_sources(lwm2m/lwm2m_device.c)
```
I think we should allow to override the default implementation for the Device and Security objects as well, i. e. introduce Kconfigs at the library level as well for these objects.

VeijoPesonen replied:
Done.

rlubos reviewed (commented):
I think this looks pretty good already. I don't have any major concerns anymore, just a few comments.

rlubos commented:
@umapraseeda An early notice, from what I understood the documentation is still not complete. Please assign some other tech writer if you won't have time to look into this PR.

tejlmand commented on lib/nrf91_lwm2m_app/Kconfig:
```diff
@@ -4,66 +4,12 @@
 #
 
-menuconfig NRF91_LWM2M_APP
+config NRF91_LWM2M_APP
```
Why is this no longer a `menuconfig` ?
Now all config settings is expanded into current menu, instead of having a menu of their own like: `LTE_LINK_CONTROL`, `NRF_MODEM_LIB`, etc. has.

VeijoPesonen replied:
Fixed.

tejlmand commented on lib/nrf91_lwm2m_app/Kconfig:
```diff
@@ -10,25 +10,28 @@ config NRF91_LWM2M_APP
+config NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT
 	bool "Connection Monitor object support"
 	default y
+	select LWM2M_CONN_MON_OBJ_SUPPORT
```
Note, if a user uses menuconfig and sets `NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT=n` then LWM2M_CONN_MON_OBJ_SUPPORT will still be enabled.

Not sure that is intended ?

Also, what is the reason for having both symbols ?
In code and CMake , it seems that only NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT is used.

See also: https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#stuck-symbols-in-menuconfig-and-guiconfig

rlubos replied:
> Also, what is the reason for having both symbols ?

`LWM2M_CONN_MON_OBJ_SUPPORT` indicates that the LwM2M library supports the Connectivity Monitoring object (supports but does not implement the object instance). 
`NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT` indicates that this library provides a default implementation of the Connectivity Monitoring object. The user should still be able to disable the default implementation and provide his own.

tejlmand replied:
> LWM2M_CONN_MON_OBJ_SUPPORT indicates that the LwM2M library supports the Connectivity Monitoring object 

So if this is something the library indicates that it supports, then it sounds like `LWM2M_CONN_MON_OBJ_SUPPORT` itself probably should be promptless ?
Someone providing an implementation, like `NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT` can then still select that symbol.
At least from your description and my interpretation, it sounds like it should not be the user who configures `LWM2M_CONN_MON_OBJ_SUPPORT`.

VeijoPesonen replied:
> Note, if a user uses menuconfig and sets `NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT=n` then LWM2M_CONN_MON_OBJ_SUPPORT will still be enabled.
... 
> See also: https://docs.zephyrproject.org/latest/guides/kconfig/tips.html#stuck-symbols-in-menuconfig-and-guiconfig

For me it looks like that issue applies only if I would do something like

```
config LWM2M_CONN_MON_OBJ_SUPPORT_FOO
	bool "Foo"

config NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT
	bool "Connection Monitor object support"
	default y if LWM2M_CONN_MON_OBJ_SUPPORT_FOO
	default n
```

But that's not the case here.

rlubos replied:
Well, I might lack Kconfig knowledge here, but I always thought that `select` guarantees that the user does not have to bother with configuring `LWM2M_CONN_MON_OBJ_SUPPORT` on his own.

`LWM2M_CONN_MON_OBJ_SUPPORT` is an upstream symbol, it should not depend on `NRF91_LWM2M_APP` library in any way. I don't know how much flexibility do we have in reconfiguring upstream Kconfig entries from `sdk-nrf` itself.

tejlmand replied:
> LWM2M_CONN_MON_OBJ_SUPPORT is an upstream symbol, 

I didn't notice that, that of course  makes it different. I thought it was a NCS defined symbol.

>  it should not depend on NRF91_LWM2M_APP library in any way. 

Agree.

> Well, I might lack Kconfig knowledge here, but I always thought that select guarantees that the user does not have to bother with configuring LWM2M_CONN_MON_OBJ_SUPPORT on his own.

Yes, as long as `NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT=y`, but if user uses `menuconfig` to disable `NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT`, then `LWM2M_CONN_MON_OBJ_SUPPORT` will still be true.

Hence the reason for initial comment.

VeijoPesonen replied:
> but if user uses menuconfig to disable NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT, then LWM2M_CONN_MON_OBJ_SUPPORT will still be true.

My interpretation is that the are no issues left anymore. Only menuconfig here should be the one which disables/enables the library.

VeijoPesonen replied:
I assume there are no objections left.

tejlmand replied:
as discussed, having LWM2M_CONN_MON_OBJ_SUPPORT=y and NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT=n which happens when user deselects NRF91_LWM2M_APP_CONN_MON_OBJ_SUPPORT should not be an issue in this case.

tejlmand reviewed (changes requested):
couple of observations

rlubos commented on include/net/nrf91_lwm2m_app.rst:
```diff
@@ -0,0 +1,195 @@
+already defined in `LwM2M Object and Resource Registry`_.
+If there is already such definition you might want to use that, instead of defining your own.
+
+Please note that Zephyr's LwM2M library has only a limited support for some IPSO objects, so extending beyound those,
```
```suggestion
Please note that Zephyr's LwM2M library has only a limited support for some IPSO objects, so extending beyond those,
```

VeijoPesonen replied:
Seems to be fixed.

rlubos commented on include/net/nrf91_lwm2m_app.rst:
```diff
@@ -0,0 +1,195 @@
+which would give us that the resource object ID would be `3300`. Within that list, we could open a `raw XML`_ file
+or open a bit more human readable view from `the LwM2M editor`_ to see what resources are defined under the object ID 3300.
+
+.. _LwM2M Object and Resource Registry: http://openmobilealliance.org/wp/OMNA/LwM2M/LwM2MRegistry.html
```
The links should rather end up in `links.txt`.

VeijoPesonen replied:
Seems to be fixed.

rlubos commented on include/net/nrf91_lwm2m_app.rst:
```diff
@@ -0,0 +1,195 @@
+Please note that Zephyr's LwM2M library has only a limited support for some IPSO objects, so extending beyound those,
+requires changes to the internal engine as well.
+
+For this example, we are going to demonstrate how do we define a resource that follow Generic Sensor definition from IPSO.
```
```suggestion
For this example, we are going to demonstrate how do we define an object that follows Generic Sensor definition from IPSO.
```
We should differentiate a resource (which reprensents an individual field in an object specification) from objects (which groups multiple resources) and do not mix these.

VeijoPesonen replied:
Seems to be fixed.

rlubos commented on include/net/nrf91_lwm2m_app.rst:
```diff
@@ -0,0 +1,195 @@
+
+For this example, we are going to demonstrate how do we define a resource that follow Generic Sensor definition from IPSO.
+First, we look for "Generic Sensor" from the `LwM2M Object and Resource Registry`_ list
+which would give us that the resource object ID would be `3300`. Within that list, we could open a `raw XML`_ file
```
```suggestion
which would give us that the object ID would be `3300`. Within that list, we could open a `raw XML`_ file
```

VeijoPesonen replied:
Seems to be fixed.

rlubos commented on include/net/nrf91_lwm2m_app.rst:
```diff
@@ -0,0 +1,195 @@
+and of course this is our first instance. Therefore our full path for the sensor value would be
+`3300/0/5700`. You need to use this path in the LwM2M library API.
+
+First thing we do, is to have read-function for our data. It is out of scope for this tutorial
```
I know it'd be hard to cover every possible use of the library, but I think it's worth mentioning that registering a read callback is optional and is recommended if we want to read the data directly from a sensor on each READ operation. 

Another common case for the readable resource is to modify their value on event, there's no need to register a read callback then. A good exaple for this is the Push Button object - whenever we get an event that the button was pressed/released we just update its value through the lwm2m_engine with `lwm2m_engine_set_bool()`. When a READ operation is issued by the server, the engine will get the button value directly from the object internal data instead of reaching the read callback.

VeijoPesonen replied:
Thanks, this got added by Uma.

rlubos reviewed (commented):
I've gone through the documentation and left a few comments. The tech writer has the final word here though.

VeijoPesonen commented:
@rlubos , @SeppoTakalo I renamed the library. Please review.

rlubos commented:
Regarding the rename, are we sure about this? The objects defined by the library are nRF91-specific (i. e. they're interfacing the modem in most cases).

SeppoTakalo commented:
> Regarding the rename, are we sure about this? The objects defined by the library are nRF91-specific (i. e. they're interfacing the modem in most cases).

But how much are they?

I mean, they are kind of generic. Even though the implementation is now nRF91 based, they could be generic. For example Firmware update. And some day in future, there would be other than nRF91 chips where this would work as well. So I'm therefore leaning towards generic name, and not pinpointing 91 series.

rlubos commented:
> > Regarding the rename, are we sure about this? The objects defined by the library are nRF91-specific (i. e. they're interfacing the modem in most cases).
> 
> But how much are they?
> 
> I mean, they are kind of generic. Even though the implementation is now nRF91 based, they could be generic. For example Firmware update. And some day in future, there would be other than nRF91 chips where this would work as well. So I'm therefore leaning towards generic name, and not pinpointing 91 series.

Well that's the point, they could be generic with some extra ifdef slicing, but they're not at this point. Connectivity Monitoring object - interfaces `modem_info` library directly, Firmware Update - allowing modem FW updates unconditionally, Location - interfacing GPS driver available only for nRF91. My point is, that we already have other chips that enable networking (for instance all of the devices that can use OpenThread) so we'll sooner or later end up with someone trying to use the library with these.

Don't get me wrong, I'm not totally against the rename, if we want to keep the library name stable for future use-cases that's fine. We should just make sure it's clear for the users of the library that is intended for nRF91 only at this point (IMO both in the documentation and the API header).

SeppoTakalo commented:
I think we are verbal enough in the documentation section. It mentions that this library works on nRF91.

But code wise, there is nothing limiting us on preparing for other uses as well. Some parts of the library would be directly usable for the OpenThread use as well.

rlubos commented:
> I think we are verbal enough in the documentation section. It mentions that this library works on nRF91.

Ok, I now see that we do mention that the library is for nRF9160-based devices, thanks.

rlubos commented on include/net/lwm2m_client_utils.rst:
```diff
@@ -0,0 +1,317 @@
+* :option:`CONFIG_LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT` enabled
+
+Support for the objects can be set individually but are enabled by default.
+Disable the :option:`CONFIG_NRF91_LWM2M_APP_DEVICE_OBJ_SUPPORT` Kconfig option only if you are implementing a ``reboot`` resource on your application because of a mandatory requirement.
```
Leftover config name.

VeijoPesonen replied:
Fixed.

rlubos commented:
This PR seems to be pretty advanced already, do you plan any more changes? Perhaps it's a good time to pull it out of the "draft" stage?

VeijoPesonen commented:
> This PR seems to be pretty advanced already, do you plan any more changes?

I don't see need for more changes when it comes to this PR.

rlubos reviewed (commented):
Looks good already since the review was mostly done in the draft stage. We should squash these commits though, we don't want the commits addressing review requests in the history. IMO a single commit would do, but you can also split the documenation to a separate one if you prefer.

VeijoPesonen commented:
> We should squash these commits though, we don't want the commits addressing review requests in the history. 

Made the split but I ended up creating three commits:
1. Implementation plus initial documentation
2. Seppo's documentation section how to create objects and resources
3. Library renaming

Please let me know if you would prefer it different way.

rlubos commented:
Actually, I don't see a reason to keep the old name in any of the commits, it'd just blur the history. So ideally we should squash the renaming changes into the other commits. But if it's a too much hassle in your opinion to separate the changes, just put everything into a single one, I don't see a problem with that.

Another thing, sorry for not spotting this earlier, we keep the networking libraries in `subsys/net/lib` directory, please move the files into that location.

Finally, we need a specific commit format in order to keep checkpatch happy, I suggest to use `net: lib: lwm2m_client_utils:`  prefix for the library (provided we move the sources).

rlubos commented:
@tejlmand Could you revisit the PR? All review comments should've been addressed already.

VeijoPesonen commented:
@rlubos, @lemrey, @SeppoTakalo I happened to stumble on the FOTA buffer alignment issue triggered by the PR #4574.
